pkg/endpoint/endpoint.go

```diff
@@ -197,36 +203,60 @@ func RunK8sCiliumEndpointSyncGC() {
-					podFullName := pod.Name + ":" + pod.Namespace
-					clusterPodSet[podFullName] = true
-				}
+				var (
+					listOpts = meta_v1.ListOptions{Limit: 10}
+					loopStop = time.Now().Add(CiliumEndpointGCInterval)
+				)
 
-				// "" is all-namespaces
-				ceps, err := ciliumClient.CiliumEndpoints(meta_v1.NamespaceAll).List(meta_v1.ListOptions{})
-				if err != nil {
-					scopedLog.WithError(err).Debug("Cannot list CEPs")
-					return err
-				}
-				for _, cep := range ceps.Items {
-					cepFullName := cep.Name + ":" + cep.Namespace
-					if _, found := clusterPodSet[cepFullName]; !found {
-						// delete
-						scopedLog = scopedLog.WithFields(logrus.Fields{
-							logfields.EndpointID: cep.Status.ID,
-							logfields.K8sPodName: cepFullName,
-						})
-						scopedLog.Debug("Orphaned CiliumEndpoint is being garbage collected")
-						if err := ciliumClient.CiliumEndpoints(cep.Namespace).Delete(cep.Name, &meta_v1.DeleteOptions{}); err != nil {
-							scopedLog.WithError(err).Debug("Unable to delete CEP")
+			perCEPFetch:
+				for time.Now().Before(loopStop) { // stop the loop if it takes too long
+					ceps, err := ciliumClient.CiliumEndpoints(meta_v1.NamespaceAll).List(listOpts)
+					switch {
+					case err != nil && k8serrors.IsResourceExpired(err) && ceps.Continue != "":
+						// This combination means we saw a 410 ResourceExpired error but we
+						// can iterate on the now-current snapshot. We need to refetch,
+						// however.
+						// See https://github.com/kubernetes/apimachinery/blob/master/pkg/apis/meta/v1/types.go#L350-L381
+						// or the docs for k8s.io/apimachinery/pkg/apis/meta/v1.ListOptions
+						// vendored into this repo.
+						listOpts.Continue = ceps.Continue
+						continue perCEPFetch
+
+					case err != nil:
+						scopedLog.WithError(err).Debug("Cannot list CEPs")
+						return err
+					}
+
+					// setup listOpts for the next iteration
+					listOpts.Continue = ceps.Continue
+
+					// For each CEP we fetched, check if we know about it
+					for _, cep := range ceps.Items {
+						cepFullName := cep.Namespace + "/" + cep.Name
+						_, exists, err := podStore.GetByKey(cepFullName)
+						if err != nil {
 							return err
 						}
+
+						if !exists {
+							// delete
+							scopedLog = scopedLog.WithFields(logrus.Fields{
+								logfields.EndpointID: cep.Status.ID,
+								logfields.K8sPodName: cepFullName,
+							})
+							scopedLog.Debug("Orphaned CiliumEndpoint is being garbage collected")
+							PropagationPolicy := meta_v1.DeletePropagationBackground // because these are const strings but the API wants pointers
+							if err := ciliumClient.CiliumEndpoints(cep.Namespace).Delete(cep.Name, &meta_v1.DeleteOptions{PropagationPolicy: &PropagationPolicy}); err != nil {
+								scopedLog.WithError(err).Debug("Unable to delete orphaned CEP")
+								return err
+							}
+						}
+					}
+					if ceps.Continue != "" {
+						// there is more data, continue
+						continue perCEPFetch
```

Comment:
Add a relaxing sleep here to pace the garbage collector.

Reply:
I added it to the start of the loop since there are two continue cases we encounter.